csrc/buffer.cpp

```diff
@@ -100,4 +100,74 @@ JBinaryBlob::~JBinaryBlob()
 
 uint8_t* JBinaryBlob::get() { return ptr_; }
 
+JIOBlobs::JIOBlobs(JNIEnv* env,
+    jobject inputDirectByteBuffer,
+    jbyteArray inputArray,
+    jobject outputDirectByteBuffer,
+    jbyteArray outputArray)
+    : env_(env)
+    , input_array_(inputArray)
+    , output_array_(outputArray)
+{
+    // First, we need to deal with buffers that are direct ByteBuffers.
+    if (inputDirectByteBuffer != nullptr) {
+        // One should be null. In the Java layer, we ensure this.
+        if (inputArray != nullptr) {
+            throw java_ex(EX_ERROR,
+                "THIS SHOULD NOT BE REACHABLE. Both inputDirectByteBuffer and inputArray cannot be provided.");
+        }
+        input_ptr_ = (uint8_t*)env->GetDirectBufferAddress(inputDirectByteBuffer);
+    }
+
+    if (outputDirectByteBuffer != nullptr) {
+        // One should be null. In the Java layer, we ensure this.
+        if (outputArray != nullptr) {
+            throw java_ex(EX_ERROR,
+                "THIS SHOULD NOT BE REACHABLE. Both outputDirectByteBuffer and outputArray cannot be provided.");
+        }
+        output_ptr_ = (inputDirectByteBuffer == outputDirectByteBuffer)
+            ? input_ptr_
+            : ((uint8_t*)env->GetDirectBufferAddress(outputDirectByteBuffer));
+    }
```

Comment:
nit: you can extract this to a method for both input/output

Reply:
They are slightly different. For the second one (in this case `outputDirectByteBuffer`), we also check if it is the same as input: if yes, we avoid calling `GetDirectBufferAddress`.